[user triton] Raise an exception when encountering nested @triton.autotune decorators or @triton.heuristics

SamGinzburg opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143519

We support running a single Autotuner for each Triton kernel. Currently,
if there are multiple autotuning decorators, the subsequent ones will be
silently ignored.

Instead, we should raise an error here to avoid silent incorrectness.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

SamGinzburg committed:
[user triton] Raise an exception when encountering nested @triton.autotune decorators

We support running a single Autotuner for each Triton kernel. Currently,
if there are multiple autotuning decorators, the subsequent ones will be
silently ignored.

Instead, we should raise an error here to avoid silent incorrectness.

[ghstack-poisoned]

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,62 @@ def f(x, increment):
 
+    @requires_gpu
+    @common_utils.parametrize("backend", ["eager", "aot_eager", "inductor"])
+    @common_utils.parametrize("autotune_at_compile_time", [True, False])
```
Remove `autotune_at_compile_time` as irrelevant?

aakhundov commented on test/inductor/test_triton_kernels.py:
```diff
@@ -3550,6 +3550,62 @@ def f(x, increment):
+            x = tl.load(x_ptr + offsets, mask=mask)
+            y = tl.load(y_ptr + offsets, mask=mask)
+            output = x + y
+            tl.atomic_add(output_ptr + offsets, output, mask=mask)
```
Nit: let's use `tl.store` here for consistency with other `add_kernel`s.

aakhundov reviewed (approved):
Thanks @SamGinzburg! Left a few minor comments.

SamGinzburg commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1179,6 +1179,21 @@ def call_triton_kernel(
+        iter_kernel = variable.kernel
+        autotuner_count = 0
+        while not isinstance(iter_kernel, JITFunction):
+            if isinstance(iter_kernel, Autotuner):
```
yeah probably, I'll add a separate test for this

SamGinzburg committed:
Update on "[user triton] Raise an exception when encountering nested @triton.autotune decorators or @triton.heuristics"


We support running a single Autotuner for each Triton kernel. Currently,
if there are multiple autotuning decorators, the subsequent ones will be
silently ignored.

Instead, we should raise an error here to avoid silent incorrectness.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg committed:
Update on "[user triton] Raise an exception when encountering nested @triton.autotune decorators or @triton.heuristics"


We support running a single Autotuner for each Triton kernel. Currently,
if there are multiple autotuning decorators, the subsequent ones will be
silently ignored.

Instead, we should raise an error here to avoid silent incorrectness.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1179,6 +1179,25 @@ def call_triton_kernel(
+                autotuner_count += 1
+            if isinstance(iter_kernel, Heuristics):
+                self.raise_unsupported(
+                    "Passing @triton.heuristics decorators is not supported. "
```
```suggestion
                    "Passing @triton.heuristics decorator after @triton.autotune decorator is not supported."
```

SamGinzburg committed:
Update on "[user triton] Raise an exception when encountering nested @triton.autotune decorators or @triton.heuristics"


We support running a single Autotuner for each Triton kernel. Currently,
if there are multiple autotuning decorators, the subsequent ones will be
silently ignored.

Instead, we should raise an error here to avoid silent incorrectness.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

SamGinzburg commented:
@pytorchbot merge
